Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp

```diff
@@ -117,10 +117,20 @@ void DNSResolveQueueCFNet::performDNSLookup(const String& hostname, Ref<Completi
     nw_parameters_set_context(parameters.get(), context.get());
     RetainPtr resolver = adoptCF(nw_resolver_create_with_endpoint(hostEndpoint.get(), parameters.get()));
 
-    nw_resolver_set_update_handler(resolver.get(), dispatch_get_main_queue(), makeBlockPtr([resolver, completionHandler = WTFMove(completionHandler)] (nw_resolver_status_t status, nw_array_t resolvedEndpoints) mutable {
+    auto timer = makeUnique<Timer>([resolver, completionHandler]() {
+        WTF::ensureOnMainThread([resolver, completionHandler]() {
+            nw_resolver_cancel(resolver.get());
+            completionHandler->complete(makeUnexpected(DNSError::Cancelled));
+        });
+    });
+    timer->startOneShot(Seconds(60));
+
+    nw_resolver_set_update_handler(resolver.get(), dispatch_get_main_queue(), makeBlockPtr([resolver, completionHandler = WTFMove(completionHandler), timer = WTFMove(timer)] (nw_resolver_status_t status, nw_array_t resolvedEndpoints) mutable {
```

Comment:
In case this got cancelled, I wonder in which thread this block is destroyed.
It should not be destroyed out of main thread, timer should be destroyed in main thread. Can you validate this?

Reply:
I think the handler is called on the main thread because of dispatch_get_main_queue() so it will always be run on the main thread and then the de-alloc will also happen on the main thread ?

Reply:
The callback will always be executed in main thread, but I am less sure about the deallocation in case the callback is not called for instance. Could you have a try by running that code with a timeout of 0 so that cancel happens? Timer or TimerBase probably have ASSERT/RELEASE_ASSERT in their destructors.

To trigger mdns resolution, you can probably use https://webrtc.github.io/samples/src/content/datachannel/basic/.